crates/types/src/replication/nodeset_selector.rs

```diff
@@ -0,0 +1,1052 @@
+        )
+        .unwrap();
+
+        assert!(nodeset.contains(1));
+        assert_eq!(nodeset.len(), 6);
+    }
+
+    #[test]
+    fn selection_with_preferred_previous_nodeset() {
+        let mut nodes_config = NodesConfiguration::default();
+        // we start by 6 nodes
+        // - region-0
+        //     zone-0  N0
+        //     zone-1  N3
+        //-  region1
+        //     zone-0  N1, N4
+        //     zone-1  N1
+        //-  region2
+        //     zone-0  N2
+        //     zone-1  N5
+        for i in 0..6 {
+            // three regions, 2 zones per region
+            let region = i % 3;
+            let zone = i % 2;
+            nodes_config.upsert_node(generate_node(
+                i,
+                StorageState::ReadWrite,
+                Role::LogServer,
+                &format!("region-{region}.zone-{zone}"),
+            ));
+        }
+        // zone:2. we have 3 * 2 = 6 zones in total.
+        let replication: ReplicationProperty = "{zone: 2}".parse().unwrap();
+
+        // we like node 1 but it's not top priority
+        let preferred_nodes = NodeSet::from_single(1);
+        let log_id = 5;
+        let options = NodeSetSelectorOptions::new(log_id).with_preferred_nodes(&preferred_nodes);
+        let nodeset1 = DomainAwareNodeSetSelector::select(
+            &nodes_config,
+            &replication,
+            logserver_candidate_filter,
+            logserver_writeable_node_filter,
+            options,
+        )
+        .unwrap();
+
+        assert_eq!(nodeset1.len(), 6);
+
+        // let's add 4 more nodes to this cluster without touching existing nodes
+        for i in 6..=10 {
+            // three regions, 2 zones per region
+            let region = i % 3;
+            let zone = i % 2;
+            nodes_config.upsert_node(generate_node(
+                i,
+                StorageState::ReadWrite,
+                Role::LogServer,
+                &format!("region-{region}.zone-{zone}"),
+            ));
+        }
+
+        // the new nodeset have the entirety of the old nodeset as preferred
+        let options = NodeSetSelectorOptions::new(log_id).with_preferred_nodes(&nodeset1);
+        let nodeset2 = DomainAwareNodeSetSelector::select(
+            &nodes_config,
+            &replication,
+            logserver_candidate_filter,
+            logserver_writeable_node_filter,
+            options,
+        )
+        .unwrap();
+
+        // nodeset2 is identical to nodeset1 because we have no reason to skip any of the
+        // previously chosen (and preferred) nodes.
+        assert_eq!(nodeset2.len(), 6);
+        assert_eq!(nodeset2, nodeset1);
+
+        // Now if we regenerate but assuming that enough nodes became unwriteable in nodeset1, we
+        // even that we have nodeset1 as preferred, we should see enough writeable nodes in
+        // nodeset2.
+        // 4 nodes. 0 became Disabled (not candidates), and the rest became readonly
+        for i in 0..5 {
```

Comment:
What does "4 nodes" refer to?

Reply:
4 nodes updated. Will update the comment.